pkg/config/draftconfig_template_test.go

```diff
@@ -108,7 +115,56 @@ func loadTemplatesWithValidation() error {
 			if _, ok := validVariableKinds[variable.Kind]; !ok {
 				return fmt.Errorf("template %s has an invalid variable kind: %s", path, variable.Kind)
 			}
+
+			// version range check once we define versions
+			// if _, err := semver.ParseRange(variable.Versions); err != nil {
+			// 	return fmt.Errorf("template %s has an invalid version range: %s", path, variable.Versions)
+			// }
+
+			allVariables[variable.Name] = variable
+			if variable.Default.ReferenceVar != "" {
+				referenceVarMap[variable.Name] = variable
+			}
+		}
+
+		for _, currVar := range referenceVarMap {
+			refVar, ok := allVariables[currVar.Default.ReferenceVar]
+			if !ok {
+				return fmt.Errorf("template %s has a variable %s with reference to a non-existent variable: %s", path, currVar.Name, currVar.Default.ReferenceVar)
+			}
+
+			if currVar.Name == refVar.Name {
+				return fmt.Errorf("template %s has a variable with cyclical reference to itself: %s", path, currVar.Name)
+			}
+
+			if isCyclicalVariableReference(currVar, refVar, allVariables, true, map[string]bool{}) {
+				return fmt.Errorf("template %s has a variable with cyclical reference to itself: %s", path, currVar.Name)
+			}
 		}
+
+		allTemplates[strings.ToLower(currTemplate.TemplateName)] = currTemplate
 		return nil
 	})
 }
+
+func isCyclicalVariableReference(initialVar, currRefVar *BuilderVar, allVariables map[string]*BuilderVar, isFirst bool, visited map[string]bool) bool {
+	if !isFirst && initialVar.Name == currRefVar.Name {
```

Comment:
what's the special case for `isFirst` for here? wouldn't that get caught by setting `visited[initialVar.Name]` to `true` and then checked later like the other variables?

Reply:
fixed